### ISSUE TITLE
Gradient of sigma w.r to x get inf values?

### ISSUE THREAD

**Issue by juuso-oskari** [ashawkey/nerf_template#5]
Hi @ashawkey and thanks for the excellent template,

I'm trying to implement NeuS to this template. But for some reason the gradients of sigma w.r to x input (needed for alpha calculations in NeuS) get inf values for some inputs and extremely high values overall. Here's how I implement the gradient calculation:
```python
def forward(self, x, direction, delta, **kwargs):
    # x: [N, 3], in [-bound, bound]
    # direction: [N, 3], normalized in [-1, 1]
    with torch.enable_grad():
        x = x.requires_grad_(True)

        sigma, feat = self.common_forward(x)

        grad = torch.autograd.grad(
                sigma,
                x,
                torch.ones_like(sigma, device=x.device),
                create_graph=True,
                retain_graph=True,
                only_inputs=True)[0]

    # alpha calculation from NeuS
    alpha = self.get_alpha(sdf=sigma, gradients=grad, dirs=direction, deltas=delta)

    # encode direction
    direction = self.view_encoder(direction)
    # get color
    color = self.view_mlp(torch.cat([feat, direction], dim=-1))
    color = torch.sigmoid(color)

    return {
        'sigma': sigma,
        'color': color,
        'gradient': grad,
        'alpha': alpha,
    }
```

And here's what I get when I print out gradient values in run_cuda():
```python
def run_cuda(self, rays_o, rays_d, bg_color=None, perturb=False, cam_near_far=None, shading='full', **kwargs):
      ....
      if self.training:
          
          xyzs, dirs, ts, rays = raymarching.march_rays_train(rays_o, rays_d, self.real_bound, self.opt.contract, self.density_bitfield, self.cascade, self.grid_size, nears, fars, perturb, self.opt.dt_gamma, self.opt.max_steps)            
          deltas = ts[:,0]
          
          dirs = dirs / torch.norm(dirs, dim=-1, keepdim=True)
          with torch.cuda.amp.autocast(enabled=self.opt.fp16):
              outputs = self(xyzs, dirs, deltas, shading=shading)
              # sigmas = outputs['sigma']
              alphas = outputs['alpha']
              print("for nan alphas: sigmas: ", outputs['sigma'][torch.isnan(alphas).squeeze()])
              print("for nan alphas: dirs: ", dirs[torch.isnan(alphas).squeeze()])
              print("for nan alphas: deltas: ", deltas[torch.isnan(alphas).squeeze()])
              print("for nan alphas: grad: ", outputs['gradient'][torch.isnan(alphas).squeeze()])
              rgbs = outputs['color']

          weights, weights_sum, depth, image = raymarching.composite_rays_train(alphas, rgbs, ts, rays, self.opt.T_thresh, True) # last true turns alpha_mode on

          results['num_points'] = xyzs.shape[0]
          results['weights'] = weights
          results['weights_sum'] = weights_sum
```
Prints out:
```python
for nan alphas: grad:  tensor([[ 14272.0000,        -inf,   9776.0000],
        [ -7408.0000,        -inf,    160.0000],
        [-15064.0000,         inf,   7172.0000],
        [ -2284.0000,   2632.0000,         inf],
        [        inf,  -6852.0000,  16592.0000],
        [ -3646.0000,        -inf,   7036.0000],
        [  5348.0000,         inf,   5116.0000],
```

**Comment by ashawkey** [comment]
@juuso-oskari Hi, actually I have done that in [nerf2mesh](https://github.com/ashawkey/nerf2mesh). You need to change the cuda kernel of gradient calculation if alpha is used as input instead of sigma: https://github.com/ashawkey/nerf2mesh/blob/main/raymarching/src/raymarching.cu#L671

**Comment by juuso-oskari** [comment]
> @juuso-oskari Hi, actually I have done that in [nerf2mesh](https://github.com/ashawkey/nerf2mesh). You need to change the cuda kernel of gradient calculation if alpha is used as input instead of sigma: https://github.com/ashawkey/nerf2mesh/blob/main/raymarching/src/raymarching.cu#L671

@ashawkey Thats very interesting, didn't know about that paper. I'm sure I can refer to it in my thesis. But isn't that raymarching implementation the same as this repository has? I guess this line should already do the change to alpha mode (from upper code)?:

````python
weights, weights_sum, depth, image = raymarching.composite_rays_train(alphas, rgbs, ts, rays, self.opt.T_thresh, True) # last true turns alpha_mode on
````

But still the gradients stay the same (containing infs)

**Comment by qingxu-thu** [comment]
I find the same problem when using the hash grid encoder, I am not sure whether somethings wrong
